### PR TITLE
[Faiss] Fix dependencies

### LINKED ISSUE #18776: [faiss] The target files on x64-linux contain absolute paths
**Describe the bug**
The target files on x64-linux contain several absolute paths to libraries. This makes it impossible to link against faiss on machines that don't have the same folder tree / installed binaries. For instance this is what I'm getting:
```
set_target_properties(faiss PROPERTIES
  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
  INTERFACE_LINK_LIBRARIES "\$<LINK_ONLY:OpenMP::OpenMP_CXX>;${_IMPORT_PREFIX}/lib/libopenblas.a;\$<LINK_ONLY:-lpthread>;${_IMPORT_PREFIX}/lib/liblapack.a;/usr/lib/x86_64-linux-gnu/libm.so;/usr/lib/x86_64-linux-gnu/libgfortran.so.4;${_IMPORT_PREFIX}/lib/libopenblas.a;\$<LINK_ONLY:-lpthread>"
)
```

**Environment**
- OS: Linux
- Compiler: x64-linux

**To Reproduce**
1. ./vcpkg install faiss --triplet=x64-linux on machine A
2. Try to statically link against faiss in a project on machine B. 
If machines A and B have different locations for where lapack & blas are installed you will get a link error.

**Expected behavior**
Linking should succeed

**Failure logs** 
-(please attached failure logs)

**Additional context**
faiss should probably just reference the lapack and BLAS::BLAS targets rather than their libraries.

### PR DESCRIPTION
**Pull request by bobuc-havok**
**Describe the pull request**

- #### What does your PR fix?  
  Fixes #18776 

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  all previously supported, No

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  Yes

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

### CODE REVIEW

File: ports/lapack-reference/portfile.cmake [hunk 1]
```diff
@@ -137,7 +137,5 @@ if(VCPKG_TARGET_IS_WINDOWS)
     endif()
 endif()
 
-if(NOT VCPKG_TARGET_IS_WINDOWS)
-    file(COPY ${CMAKE_CURRENT_LIST_DIR}/vcpkg-cmake-wrapper.cmake DESTINATION ${CURRENT_PACKAGES_DIR}/share/lapack)
-    file(COPY ${CMAKE_CURRENT_LIST_DIR}/FindLAPACK.cmake DESTINATION ${CURRENT_PACKAGES_DIR}/share/lapack)
-endif()
+file(COPY ${CMAKE_CURRENT_LIST_DIR}/vcpkg-cmake-wrapper.cmake DESTINATION ${CURRENT_PACKAGES_DIR}/share/lapack)
```
Comment (NancyLi1013) [inline ports/lapack-reference/portfile.cmake:140]: Why do you remove the condition for these?

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Comment by NancyLi1013** [comment]
`faiss `failed on x64-windows with the following errors:
```
/pdb:faiss\faiss.pdb /dll /version:0.0 /machine:x64 /nologo /debug /INCREMENTAL /DEF:faiss\CMakeFiles\faiss.dir\.\exports.def lapack.lib D:\installed\x64-windows\debug\lib\openblas.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:faiss\CMakeFiles\faiss.dir/intermediate.manifest faiss\CMakeFiles\faiss.dir/manifest.res" failed (exit code 1104) with the following output:
LINK : fatal error LNK1104: cannot open file 'lapack.lib'
ninja: build stopped: subcommand failed.
```

File: ports/lapack-reference/portfile.cmake (hunk 1 above)
Comment (bucurb) [reply ports/lapack-reference/portfile.cmake:140]: Well, it looked like a bug to me. Both clapack and lapack-reference are implementations of the 'pure virtual' lapack package. Basically you can just vcpkg install lapack and based on the triplet you get one of the 2 implementations. But lapack gets to decide which implementation goes on which platform. If you have a look at the clapack portfile it has similar logic whereby it copeis the FindLapack.cmake files into lapack, but it does so all the time, regardless of the platform. Lapack-reference had this additional platform check which did not even match the one in lapack. So for x64-windows you would get lapack -> lapack-reference but there would be no FindLapack.cmake installed in lapack\.

**Comment by bucurb** [comment]
@NancyLi1013 yup, saw that build break, it went green now :)

**Review by NancyLi1013 — APPROVED** [review]

**Comment by NancyLi1013** [comment]
LGT now, have you tested the feature `gpu`?

**Comment by bucurb** [comment]
@NancyLi1013 good point, I added a small fix to the port file for gpu. That said faiss[gpu] does not build for me locally, regardless of any changes. I'm getting nvcc compiler errors such as:

```
faiss/gpu/utils/Tensor.cuh(366): error: argument list for template "T::canUseIndexType" is missing
faiss/gpu/utils/PtxUtils.cuh(25): error: asm operand type size(4) does not match type/size implied by constraint 'l'
```
There are others. In terms of CMake targets, the gpu feature should be fine, as the dependency is declared as:

```
find_package(CUDAToolkit REQUIRED)
target_link_libraries(faiss PRIVATE CUDA::cudart CUDA::cublas)
```

so it should not generate absolute paths. I'm not sure if the CI builds test faiss[gpu], but in case the failure to compile the cuda code is more general than on my machine, I think it should be split out from this PR / issue.

**Comment by NancyLi1013** [comment]
`gpu `is not default feature, so it will not be tested on CI. Thanks for your fix and information. Now can you build faiss[gpu] on your local machine?

**Comment by bucurb** [comment]
@NancyLi1013 no I cannot build `faiss[gpu]:x64-windows` locally, I'm getting the nvcc compiler errors mentioned above. It might be because I don't have an NVidia graphics card on my machine, not sure.

**Comment by NancyLi1013** [comment]
I was blocked by the failure on my local with master branch.
```
CMake Error at CMakeLists.txt:26 (enable_language):
  No CMAKE_CUDA_COMPILER could be found.
```
Have you tried with master branch on your machine? Does it work? If you have the same problem with the changes on this PR, it might not be the new issue introduced by this PR.

**Comment by bucurb** [comment]
@NancyLi1013 yup, I am getting the same cuda compiler not found error on master. Seems to be common, as other ports have fixes for it. It looks like you can't get to compiling cuda on master until you make changes and fix the cuda compiler like I have in the portfile in this PR. I don't think my changes broke the faiss[cuda] - I think it was simply broken from the start and not tested on CI.

**Comment by NancyLi1013** [comment]
Thanks for your efforts and patience. @bucurb

**Comment by vicroms** [comment]
Thanks for the PR!

**Pull request merged**